src/api/app/components/chart_component.rb

```diff
@@ -0,0 +1,102 @@
+class ChartComponent < ApplicationComponent
+  attr_reader :actions
+
+  def initialize(actions:)
+    super
+
+    @actions = actions
+  end
+
+  def build_results_data
+    raw_data = []
+
+    # take all the build results for all the actions (only actions where it makes sense to have a build status)
+    @actions.where(type: [:submit, :maintenance_incident, :maintenance_release]).each do |action|
+      bs_request = BsRequest.find(action.bs_request_id)
+      # consider staging project
+      prj_name = bs_request.staging_project_id.nil? ? action.source_project : bs_request.staging_project.name
+      src_prj_obj = Project.find_by_name(prj_name)
+      src_pkg_obj = Package.find_by_project_and_name(src_prj_obj.name, action.source_package)
+
+      # the package might not exist yet in the staging project, for instance
+      next unless src_pkg_obj
+
+      # fetch all build results for the source package (considering multibuild packages as well)
+      action_build_results = src_pkg_obj.buildresult(src_prj_obj, show_all: true).results.flat_map { |_k, v| v }
+
+      next unless action_build_results
+
+      # carry over and expose relevant information
+      raw_data = raw_data.push(
+        action_build_results.map do |result_entry|
+          {
+            architecture: result_entry.architecture,
+            repository: result_entry.repository,
+            status: result_entry.code,
+            package_name: src_pkg_obj.name,
+            project_name: src_prj_obj.name
+          }
+        end
+      ).flatten
+    end
+
+    raw_data
```

Comment:
```suggestion
      raw_data.push(
        action_build_results.map do |result_entry|
          {
            architecture: result_entry.architecture,
            repository: result_entry.repository,
            status: result_entry.code,
            package_name: src_pkg_obj.name,
            project_name: src_prj_obj.name
          }
        end
      )
    end

    raw_data.flatten
```

Reply:
@ncounter I think calling `flatten` once after everything is collected in the `raw_data` array would be enough right?

Reply:
That is true, fixing it :+1: thanks